### ISSUE TITLE
[opencl] Convolution tests fail

### ISSUE THREAD

**Issue by woutdenolf** [silx-kit/silx#3094]
I disabled my intel drivers (which we have in CI AFAIK) and I only have the nvidia drivers enabled (debian10, python 3.7.3):

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 390.116                Driver Version: 390.116                   |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 580     Off  | 00000000:17:00.0 N/A |                  N/A |
| 41%   46C   P12    N/A /  N/A |    137MiB /  1474MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
|   1  Quadro P400         Off  | 00000000:73:00.0 Off |                  N/A |
| 34%   34C    P8    N/A /  N/A |     17MiB /  1999MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
```

Some kernels do not compile (something to do with textures according to @kif ). I'm only showing one of them:

```python
python run_tests.py silx.opencl.test.test_convolution.test_convolution

ERROR: test_1D (silx.opencl.test.test_convolution.TestConvolution)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mntdirect/_data_id21_inhouse/wout/dev/silx/build/lib.linux-x86_64-3.7/silx/opencl/test/test_convolution.py", line 198, in test_1D
    self.template_test("test_1D")
  File "/mntdirect/_data_id21_inhouse/wout/dev/silx/build/lib.linux-x86_64-3.7/silx/opencl/test/test_convolution.py", line 176, in template_test
    conv = self.instantiate_convol(data.shape, kernel)
  File "/mntdirect/_data_id21_inhouse/wout/dev/silx/build/lib.linux-x86_64-3.7/silx/opencl/test/test_convolution.py", line 126, in instantiate_convol
    extra_options={"dont_use_textures": not(self.param["use_textures"])}
  File "/mntdirect/_data_id21_inhouse/wout/dev/silx/build/lib.linux-x86_64-3.7/silx/opencl/convolution.py", line 83, in __init__
    self._init_kernels()
  File "/mntdirect/_data_id21_inhouse/wout/dev/silx/build/lib.linux-x86_64-3.7/silx/opencl/convolution.py", line 245, in _init_kernels
    compile_options=compile_options
  File "/mntdirect/_data_id21_inhouse/wout/dev/silx/build/lib.linux-x86_64-3.7/silx/opencl/processing.py", line 265, in compile_kernels
    self.program = pyopencl.Program(self.ctx, kernel_src).build(options=compile_options)
  File "/mntdirect/_data_id21_inhouse/wout/dev/virtualenvs/lid21data/silx/py37/lib/python3.7/site-packages/pyopencl/__init__.py", line 522, in build
    options_bytes=options_bytes, source=self._source)
  File "/mntdirect/_data_id21_inhouse/wout/dev/virtualenvs/lid21data/silx/py37/lib/python3.7/site-packages/pyopencl/__init__.py", line 566, in _build_and_catch_errors
    raise err
pyopencl._cl.RuntimeError: clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE

Build on <pyopencl.Device 'GeForce GTX 580' on 'NVIDIA CUDA' at 0x15a7b30>:

<kernel>:173:31: error: no matching function for call to 'read_imagef'
        sum += READ_IMAGE_X * READ_FILTER_VAL(jx);
                              ^~~~~~~~~~~~~~~~~~~
<kernel>:99:32: note: expanded from macro 'READ_FILTER_VAL'
    #define READ_FILTER_VAL(j) READ_FILTER_1D(j)
                               ^~~~~~~~~~~~~~~~~
<kernel>:48:27: note: expanded from macro 'READ_FILTER_1D'
#define READ_FILTER_1D(j) read_imagef(filter, (int2) (FILTER_INDEX(j), 0)).x;
                          ^~~~~~~~~~~
cl_kernel.h:18365:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image3d_t image, sampler_t sampler, float4 coord);
                        ^
cl_kernel.h:18364:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image3d_t image, sampler_t sampler, int4 coord);
                        ^
cl_kernel.h:18363:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image2d_t image, sampler_t sampler, float2 coord);
                        ^
cl_kernel.h:18362:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image2d_t image, sampler_t sampler, int2 coord);
                        ^

(options: -DUSED_CONV_MODE=0 -DIMAGE_DIMS=1 -DFILTER_DIMS=1 -I /mntdirect/_data_id21_inhouse/wout/dev/virtualenvs/lid21data/silx/py37/lib/python3.7/site-packages/pyopencl/cl)
(source saved as /tmp/tmpik4ni302.cl)

======================================================================
SKIPPED: test_1D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_batched_2D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_nonseparable_2D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_nonseparable_3D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_separable_2D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_separable_3D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_1D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_batched_2D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_nonseparable_2D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_nonseparable_3D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_separable_2D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_separable_3D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_1D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_batched_2D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_nonseparable_2D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_nonseparable_3D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_separable_2D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_separable_3D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_1D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_batched_2D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_nonseparable_2D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_nonseparable_3D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_separable_2D (silx.opencl.test.test_convolution.TestConvolution)
SKIPPED: test_separable_3D (silx.opencl.test.test_convolution.TestConvolution)
----------------------------------------------------------------------
mode=constant not implemented without textures
----------------------------------------------------------------------
Ran 192 tests in 8.292s

FAILED (errors=96, skipped=24)
```

**Comment by pierrepaleo** [comment]
The problem is likely the very old GPU used (GTX 580, Fermi generation) where the driver has not the same API for `read_imagef`.

This can be fixed by several ways:
  - Not using this GPU: either via `CUDA_VISIBLE_DEVICES` or `PYOPENCL_CTX` environment variables, or with `Convolution(..., ctx=my_ctx)` where `my_ctx` is a context created on another device
  - Using `Convolution(..., extra_options={"dont_use_textures": True})`
  - By detecting the GPU generation (`C.ctx.devices[0].compute_capability_major_nv`) and disabling textures if needed

The question is, do we want to support the Fermi generation ?

**Comment by woutdenolf** [comment]
Is it feasible to have silx select a different GPU based on `ctx.devices[0].compute_capability_major_nv` before trying to compile these kernels? Or is that a never ending story? On this machine the tests run fine on the other GPU.

**Comment by pierrepaleo** [comment]
Theoretically silx already picks automatically the "best" device when instantiating an `OpenclProcessing` class. However the choice is not based on the "compute capability", but on a heuristic on the device computing power (which is more versatile). 

I don't think the "pick by compute capability" should be included in silx, since we try to be not nvidia-centric. 

That being said, the question about supporting the Fermi generation remains. I have an opinion, but let's decide together.

**Comment by t20100** [comment]
Dropping Fermi means picking by compute capability right?

**Comment by pierrepaleo** [comment]
For me it means "no code dedicated to Fermi, assume that we don't use a Fermi device" (as "no code dedicated to python 2, assume we don't use "python 2").

**Comment by t20100** [comment]
OK, and same as for Python2, no `raise NoneSenseError`, sounds fair.

**Comment by pierrepaleo** [comment]
:grin:
The problem might have another cause than initially thought:
  - @woutdenolf uses driver 390.116 and experiences the problem
  - @kif uses driver 390.132 and does not have any issue

So perhaps just updating the driver will fix the issue.

Thanks @kif for spotting this out

**Comment by woutdenolf** [comment]
I upgraded the driver but still the same issue.

```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 390.132                Driver Version: 390.132                   |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 580     Off  | 00000000:17:00.0 N/A |                  N/A |
| 43%   50C   P12    N/A /  N/A |      1MiB /  1474MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
|   1  Quadro P400         Off  | 00000000:73:00.0 Off |                  N/A |
| 34%   34C    P8    N/A /  N/A |     17MiB /  1999MiB |      0%      Default |
+-------------------------------+----------------------+----------------------+
```

**Comment by kif** [comment]
Sorry, by explicitely targetting the Fermi-GPU I manage to reproduce the bug:

CUDA_VISIBLE_DEVICES=1 python3 run_tests.py silx.opencl.test.suite

======================================================================
ERROR: test_separable_3D (silx.opencl.test.test_convolution.TestConvolution)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jerome/workspace/silx/build/lib.linux-x86_64-3.7/silx/opencl/test/test_convolution.py", line 204, in test_separable_3D
    self.template_test("test_separable_3D")
  File "/home/jerome/workspace/silx/build/lib.linux-x86_64-3.7/silx/opencl/test/test_convolution.py", line 176, in template_test
    conv = self.instantiate_convol(data.shape, kernel)
  File "/home/jerome/workspace/silx/build/lib.linux-x86_64-3.7/silx/opencl/test/test_convolution.py", line 126, in instantiate_convol
    extra_options={"dont_use_textures": not(self.param["use_textures"])}
  File "/home/jerome/workspace/silx/build/lib.linux-x86_64-3.7/silx/opencl/convolution.py", line 83, in __init__
    self._init_kernels()
  File "/home/jerome/workspace/silx/build/lib.linux-x86_64-3.7/silx/opencl/convolution.py", line 245, in _init_kernels
    compile_options=compile_options
  File "/home/jerome/workspace/silx/build/lib.linux-x86_64-3.7/silx/opencl/processing.py", line 265, in compile_kernels
    self.program = pyopencl.Program(self.ctx, kernel_src).build(options=compile_options)
  File "/usr/lib/python3/dist-packages/pyopencl/__init__.py", line 510, in build
    options_bytes=options_bytes, source=self._source)
  File "/usr/lib/python3/dist-packages/pyopencl/__init__.py", line 554, in _build_and_catch_errors
    raise err
pyopencl._cl.RuntimeError: clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE - clBuildProgram failed: BUILD_PROGRAM_FAILURE

Build on <pyopencl.Device 'NVS 310' on 'NVIDIA CUDA' at 0x2649bd0>:

<kernel>:173:31: error: no matching function for call to 'read_imagef'
        sum += READ_IMAGE_X * READ_FILTER_VAL(jx);
                              ^~~~~~~~~~~~~~~~~~~
<kernel>:99:32: note: expanded from macro 'READ_FILTER_VAL'
    #define READ_FILTER_VAL(j) READ_FILTER_1D(j)
                               ^~~~~~~~~~~~~~~~~
<kernel>:48:27: note: expanded from macro 'READ_FILTER_1D'
#define READ_FILTER_1D(j) read_imagef(filter, (int2) (FILTER_INDEX(j), 0)).x;
                          ^~~~~~~~~~~
cl_kernel.h:18365:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image3d_t image, sampler_t sampler, float4 coord);
                        ^
cl_kernel.h:18364:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image3d_t image, sampler_t sampler, int4 coord);
                        ^
cl_kernel.h:18363:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image2d_t image, sampler_t sampler, float2 coord);
                        ^
cl_kernel.h:18362:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image2d_t image, sampler_t sampler, int2 coord);
                        ^
<kernel>:202:31: error: no matching function for call to 'read_imagef'
        sum += READ_IMAGE_Y * READ_FILTER_VAL(jy);
                              ^~~~~~~~~~~~~~~~~~~
<kernel>:99:32: note: expanded from macro 'READ_FILTER_VAL'
    #define READ_FILTER_VAL(j) READ_FILTER_1D(j)
                               ^~~~~~~~~~~~~~~~~
<kernel>:48:27: note: expanded from macro 'READ_FILTER_1D'
#define READ_FILTER_1D(j) read_imagef(filter, (int2) (FILTER_INDEX(j), 0)).x;
                          ^~~~~~~~~~~
cl_kernel.h:18365:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image3d_t image, sampler_t sampler, float4 coord);
                        ^
cl_kernel.h:18364:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image3d_t image, sampler_t sampler, int4 coord);
                        ^
cl_kernel.h:18363:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image2d_t image, sampler_t sampler, float2 coord);
                        ^
cl_kernel.h:18362:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image2d_t image, sampler_t sampler, int2 coord);
                        ^
<kernel>:231:31: error: no matching function for call to 'read_imagef'
        sum += READ_IMAGE_Z * READ_FILTER_VAL(jz);
                              ^~~~~~~~~~~~~~~~~~~
<kernel>:99:32: note: expanded from macro 'READ_FILTER_VAL'
    #define READ_FILTER_VAL(j) READ_FILTER_1D(j)
                               ^~~~~~~~~~~~~~~~~
<kernel>:48:27: note: expanded from macro 'READ_FILTER_1D'
#define READ_FILTER_1D(j) read_imagef(filter, (int2) (FILTER_INDEX(j), 0)).x;
                          ^~~~~~~~~~~
cl_kernel.h:18365:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image3d_t image, sampler_t sampler, float4 coord);
                        ^
cl_kernel.h:18364:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image3d_t image, sampler_t sampler, int4 coord);
                        ^
cl_kernel.h:18363:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image2d_t image, sampler_t sampler, float2 coord);
                        ^
cl_kernel.h:18362:25: note: candidate function not viable: requires 3 arguments, but 2 were provided
float4 __OVERLOADABLE__ read_imagef(image2d_t image, sampler_t sampler, int2 coord);
                        ^

(options: -DUSED_CONV_MODE=3 -DIMAGE_DIMS=3 -DFILTER_DIMS=1 -I /usr/lib/python3/dist-packages/pyopencl/cl)
(source saved as /tmp/tmpssv0tta0.cl)

======================================================================

**Comment by t20100** [comment]
closed by #3101 on branch 0.13
